File: src/my_project/interviews/top_150_questions_round_22/ex_94_design_add_and_search_words_data_structure.py
from typing import List, Union, Collection, Mapping, Optional
from abc import ABC, abstractmethod
from collections import deque, defaultdict

class WordDictionary:

    def __init__(self):
        self.root = {}
        

    def addWord(self, word: str) -> None:
        node = self.root
        for char in word:
            if char not in node:
                node[char] = {}
            node = node[char]
        node['$'] = True  # Mark end of word
        

    def search(self, word: str) -> bool:
        def dfs(node, i):
            if i == len(word):
                return '$' in node
            
            char = word[i]
            if char == '.':
                # Wildcard: try all possible characters at this position
                for key in node:
                    if key != '$' and dfs(node[key], i + 1):
                        return True
                return False
            else:
                # Exact character match
                if char not in node:
                    return False
                return dfs(node[char], i + 1)
        
        return dfs(self.root, 0)


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

File: src/my_project/interviews_typescript/top_150_questions_round_1/ex_94_design_add_and_search_words_data_structure.ts
class WordDictionary {
    private root: Map<string, any>;

    constructor() {
        this.root = new Map();
    }

    addWord(word: string): void {
        let node = this.root;
        for (const char of word) {
            if (!node.has(char)) {
                node.set(char, new Map());
            }
            node = node.get(char);
        }
        node.set('$', true);  // Mark end of word
    }

    search(word: string): boolean {
        const dfs = (node: Map<string, any>, i: number): boolean => {
            if (i === word.length) {
                return node.has('$');
            }
            
            const char = word[i];
            if (char === '.') {
                // Wildcard: try all possible characters at this position
                for (const [key, childNode] of node.entries()) {
                    if (key !== '$' && dfs(childNode, i + 1)) {
                        return true;
                    }
                }
                return false;
            } else {
                // Exact character match
                if (!node.has(char)) {
                    return false;
                }
                return dfs(node.get(char), i + 1);
            }
        };
        
        return dfs(this.root, 0);
    }
}